File: week6/kotlin/method_chaining_pattern/builder_pattern/Person.kt
class Person {
    var name: String? = null
    var age: Int? = null
    var address: String? = null
}

class PersonBuilder {
    private val person = Person()

    fun withName(name: String): PersonBuilder {
        person.name = name
        return this
    }

    fun withAge(age: Int): PersonBuilder {
        person.age = age
        return this
    }

    fun withAddress(address: String): PersonBuilder {
        person.address = address
        return this
    }

    fun build(): Person {
        return person
    }
}

fun main() {
    val person = PersonBuilder()
        .withName("홍길동")
        .withAge(20)
        .withAddress("양산시 부산대학로")
        .build()

    println("Name: ${person.name}")
    println("Age: ${person.age}")
    println("Address: ${person.address}")
}

File: week6/cpp/method_chaining_pattern/builder_pattern/Person.cpp
#include "Person.hpp"

Person::Person() : name(""), age(0), address("") {}

PersonBuilder& PersonBuilder::withName(const std::string& name) {
    person.name = name;
    return *this;
}

PersonBuilder& PersonBuilder::withAge(int age) {
    person.age = age;
    return *this;
}

PersonBuilder& PersonBuilder::withAddress(const std::string& address) {
    person.address = address;
    return *this;
}

Person PersonBuilder::build() {
    return person;
}
